File: marketdna-backend/app/models/delivery.py
"""Pydantic models for Delivery Percentage Intelligence."""
from typing import Optional
from pydantic import BaseModel


class SignalOccurrence(BaseModel):
    date: str
    ret_1d: float
    ret_1w: float
    ret_1m: float
    ret_1y: Optional[float] = None   # null when < 252 bars of forward data exist


class DeliverySignalStats(BaseModel):
    occurrences: int
    avg_1d: float
    avg_1w: float
    avg_1m: float
    win_rate: float
    expected_value: float
    max_adverse: float
    distribution: dict[str, int]
    timeline: list[SignalOccurrence]
    # Market regime conditional
    mkt_regime_high_win_rate: Optional[float] = None
    mkt_regime_low_win_rate:  Optional[float] = None
    mkt_regime_high_ev:       Optional[float] = None
    mkt_regime_low_ev:        Optional[float] = None
    mkt_regime_high_occurrences: int = 0
    mkt_regime_low_occurrences:  int = 0


class DeliverySignal(BaseModel):
    name: str
    description: str
    direction: str                  # "Bullish" | "Bearish" | "Mixed"
    is_active: bool
    current_delivery_pct: float
    current_vol_ratio: float
    grade: str
    edge_score: float
    stats: DeliverySignalStats


class DeliveryBar(BaseModel):
    date: str
    close: float
    volume: float
    delivery_pct: float
    vol_ratio: float
    delivery_spike: bool


class DeliveryReport(BaseModel):
    symbol: str
    data_start: str
    data_end: str
    total_bars: int
    current_delivery_pct: float
    avg_delivery_pct_20d: float
    signals: list[DeliverySignal]
    history: list[DeliveryBar]
    computed_at: str


class DeliverySummaryItem(BaseModel):
    symbol: str
    current_delivery_pct: float
    avg_delivery_pct_20d: float
    delivery_ratio: float
    current_vol_ratio: float          # today's volume / 20D avg volume
    best_signal: str
    grade: str
    edge_score: float
    win_rate: float
    is_active: bool
    mkt_regime_high_win_rate: Optional[float] = None
    mkt_regime_low_win_rate:  Optional[float] = None


class DeliverySummaryResult(BaseModel):
    items: list[DeliverySummaryItem]
    total_symbols: int
    computed_at: str

File: marketdna-web/src/types/delivery.ts
export interface SignalOccurrence {
  date: string
  ret_1d: number
  ret_1w: number
  ret_1m: number
  ret_1y: number | null
}

export interface DeliverySignalStats {
  occurrences: number
  avg_1d: number
  avg_1w: number
  avg_1m: number
  win_rate: number
  expected_value: number
  max_adverse: number
  distribution: Record<string, number>
  timeline: SignalOccurrence[]
  mkt_regime_high_win_rate: number | null
  mkt_regime_low_win_rate: number | null
  mkt_regime_high_ev: number | null
  mkt_regime_low_ev: number | null
  mkt_regime_high_occurrences: number
  mkt_regime_low_occurrences: number
}

export interface DeliverySignal {
  name: string
  description: string
  direction: string
  is_active: boolean
  current_delivery_pct: number
  current_vol_ratio: number
  grade: string
  edge_score: number
  stats: DeliverySignalStats
}

export interface DeliveryBar {
  date: string
  close: number
  volume: number
  delivery_pct: number
  vol_ratio: number
  delivery_spike: boolean
}

export interface DeliveryReport {
  symbol: string
  data_start: string
  data_end: string
  total_bars: number
  current_delivery_pct: number
  avg_delivery_pct_20d: number
  signals: DeliverySignal[]
  history: DeliveryBar[]
  computed_at: string
}

export interface DeliverySummaryItem {
  symbol: string
  current_delivery_pct: number
  avg_delivery_pct_20d: number
  delivery_ratio: number
  current_vol_ratio: number
  best_signal: string
  grade: string
  edge_score: number
  win_rate: number
  is_active: boolean
  mkt_regime_high_win_rate: number | null
  mkt_regime_low_win_rate: number | null
}

export interface DeliverySummaryResult {
  items: DeliverySummaryItem[]
  total_symbols: number
  computed_at: string
}

export const GRADE_COLOR: Record<string, string> = {
  'A+': '#22c55e', 'A': '#86efac', 'B': '#fbbf24', 'C': '#94a3b8', 'D': '#ef4444',
}

export const DIR_COLOR: Record<string, string> = {
  Bullish: '#22c55e', Bearish: '#ef4444', Mixed: '#fbbf24',
}

export const DIST_BUCKET_ORDER = ['<-10', '-10:-5', '-5:0', '0:5', '5:10', '>10']
